kart/spatial_tree.py

```diff
@@ -488,64 +396,127 @@ def _find_geometry_column(fields):
+    # precision if we gave them a valid range of only 360 degrees and stored max-x as an offset from min-x, but
+    # it makes the code more complicated.
+    result = _encode_value(envelope[0], -180, 540, math.floor)
+    result <<= BITS_PER_VALUE
+    result |= _encode_value(envelope[1], -180, 540, math.ceil)
+    result <<= BITS_PER_VALUE
+    result |= _encode_value(envelope[2], -90, 90, math.floor)
+    result <<= BITS_PER_VALUE
+    result |= _encode_value(envelope[3], -90, 90, math.ceil)
+    assert 0 <= result <= MAX_ENCODED_ENVELOPE
+    return result.to_bytes(BYTES_PER_ENVELOPE, BYTE_ORDER)
+
+
+def _encode_value(value, min_value, max_value, round_fn):
+    normalised = (value - min_value) / (max_value - min_value)
+    encoded = round_fn(normalised * MAX_ENCODED_VALUE)
+    assert 0 <= encoded <= MAX_ENCODED_VALUE
+    return encoded
+
+
+def decode_envelope(encoded):
+    """Inverse of encode_envelope."""
+    encoded = int.from_bytes(encoded, BYTE_ORDER)
+    assert 0 <= encoded <= MAX_ENCODED_ENVELOPE
+    max_y = _decode_value(encoded & MAX_ENCODED_VALUE, -90, 90)
+    encoded >>= BITS_PER_VALUE
+    min_y = _decode_value(encoded & MAX_ENCODED_VALUE, -90, 90)
+    encoded >>= BITS_PER_VALUE
+    max_x = _decode_value(encoded & MAX_ENCODED_VALUE, -180, 540)
+    encoded >>= BITS_PER_VALUE
+    min_x = _decode_value(encoded & MAX_ENCODED_VALUE, -180, 540)
+    return min_x, max_x, min_y, max_y
+
+
+def _decode_value(encoded, min_value, max_value):
+    assert 0 <= encoded <= MAX_ENCODED_VALUE
+    normalised = encoded / MAX_ENCODED_VALUE
+    return normalised * (max_value - min_value) + min_value
 
 
 INF = float("inf")
 
 
 def get_envelope_for_indexing(geom, transforms):
+    """
+    Returns an envelope in EPSG:4326 that contains the entire geometry.
+    Uses any and all of the given transforms to convert to EPSG:4326 (this is so that this index
+    will return the right results even if the CRS of this feature has changed at some point).
+    The returned envelope is ordered according to the GPKG specification: min-x, max-x, min-y, max-y.
+    """
+
     # TODO: Make this more rigorous - ie, make sure that:
     # -90 <= S <= N <= +90
     # -180 <= W <= +180
     # 0 <= (E - W) <= 360 ; therefore -180 <= W <= +540
 
-    # Result initialised to an infinite and negative area which disappears when unioned with.
-    result = [+INF, +INF, -INF, -INF]
+    # Result initialised to an infinite and negative area which disappears when unioned.
+    result = [+INF, -INF, +INF, -INF]
 
-    # FIXME: For correctness, this should transform the original geometries before taking their bounding box.
-    # The downside of that approach is that it is maybe 10x slower.
-    w, e, s, n = geom.envelope(only_2d=True, calculate_if_missing=True)
-    raw_envelope = w, s, e, n
+    raw_envelope = geom.envelope(only_2d=True, calculate_if_missing=True)
```

Comment:
this was west-east, now it's min-x, max-x...? They're not the same thing.

Reply:
Reading more: in the [GeoPackage spec](https://www.geopackage.org/spec130/#flags_layout) they are minx,maxx... 😡

Reply:
here is probably where if we're working with a geographic CRS we need to figure out whether our geometry spans the anti-meridian, and then calculate a proper envelope.

Reply:
Yeah in the system I was planning to use with the range -180 to 540, such that it is always W <= E - in that case, the minimum x value and the westernmost X value are the same thing. So within that system, I'm free to mix up the names min-x and west if I decide to be confusing.
However, I'm still not really sure what kind of values I might be getting back from GPKG / OGR as the min-x and max-x values, so that's the thing I need to look into and fix up next.